### DISCUSSION TITLE
Doesn't seem to read the correct .env.${NODE_ENV} file

### DISCUSSION THREAD

**Discussion by PrimeDominus** [rubiin/ultimate-nest#86] [Announcements]
This looks like an awesome template repo. A colleague and I spent quite a bit of time on this, but even though we created a `.env.dev` file in the `env` directory (hence `./env/.env.dev`) it cannot read from this file. For example, we tried first running:

`NODE_ENV=dev make migrate`

But, the mikro-orm-cli would ask for the dbName or clientUrl values, meaning it couldn't read them from `./env/.env.dev`, 

![image](https://user-images.githubusercontent.com/31121241/187201538-6ec2ae05-0781-4b54-84ed-94b669f44d37.png)

We're sure of this because we then copied `./env/.env.dev` to the project root and renamed to `.env` and once again ran `NODE_ENV=dev make migrate` which works and the db migration completes successfully.

Furthermore, after this we removed the `.env` we created from the root project folder and ran `NODE_ENV=dev npm run start:dev` and we got the below error stating the REDIS_TTL is required, which is indeed correctly set in our `./env/.env.dev` but it seems this file is not being read correctly.

![image](https://user-images.githubusercontent.com/31121241/187201443-f3372c8f-1af3-4f15-9f93-fccbebb58dcc.png)

Any advice would be appreciated 😄 

Below is our `./env/.env.dev` file
```

### Environment File ##################################################################
# This file is a "template" of which env vars need to be defined for your application
########################################################################################
	
APP_PORT=8000
APP_PREFIX=v1
NODE_ENV=dev
API_URL=http://localhost:8000
SWAGGER_USER=cit-dev
SWAGGER_PASSWORD=xW4oH6*fd3u^vi
	
	# database 
DB_HOST=localhost
DB_PORT=5432
DB_USERNAME=root
DB_PASSWORD=root
DB_DATABASE=ultimateNest


	# for encryption,decryption stuffs all over the app

ENC_IV=A46B761159225E47
ENC_KEY=EC77F2C6EE0F3611A988921B65CA8776

  # jwt
  # access expiry can be number or 35d

JWT_ACCESS_EXPIRY=365d
JWT_REFRESH_EXPIRY=999999
JWT_SECRET=EC77F2C6EE0F36


	# mail
	# template dir should start from src, mail server values smtp,ses
MAIL_HOST=smtp.gmail.com
MAIL_PASSWORD=password
MAIL_PORT=465
MAIL_SERVER=SMTP
MAIL_PREVIEW_EMAIL=false
MAIL_SENDER_EMAIL=info@exmaple.com
MAIL_TEMPLATE_DIR=resources/templates
MAIL_USERNAME=username

	#cloudinary
CLOUDINARY_CLOUD_NAME=your-cloudinary-cloud-name
CLOUDINARY_API_KEY=your-cloudinary-api-key
CLOUDINARY_API_SECRET=your-cloudinary-secret

  #redis
REDIS_URI=redis://default:test@1234@redis:6379
REDIS_TTl=10

  #rabbitmq
RABBITMQ_URI=
RABBITMQ_EXCHANGE=

  #sentry
SENTRY_DSN=https://redacted

  #google ouath
GOOGLE_CLIENT_ID=
GOOGLE_SECRET=
GOOGLE_CALLBACK_URL=http://localhost:8000/v1/auth/google/redirect


  #facebook ouath
FACEBOOK_CLIENT_ID=
FACEBOOK_CLIENT_SECRET=
FACEBOOK_CALLBACK_URL=http://localhost:8000/v1/auth/facebook/redirect

```

**Comment by rubiin** [comment]
First of all, sorry for the delay in responding. Must have missed the notification for the issue. Let me try to debug it and come back to you

**Comment by rubiin** [comment]
Tried replicating but worked successfully. Check the screenshots too
![Screenshot](https://i.imgur.com/tYDnLot.png)

![Screenshot](https://i.imgur.com/4avmeZy.png)

**Comment by rubiin** [comment]
Also, I have added some env fixes so could you try on the latest pull

**Comment by PrimeDominus** [comment]
We tried the latest on Fedora and Windows using WSL and we had the same issue.

**Comment by rubiin** [comment]
Works as expected. Can you guys maybe share the steps you did.  Also on config module can you directly pass the file name on the .forRoot https://github.com/rubiin/ultimate-nest/blob/master/src/lib/config/config.module.ts

**Comment by rubiin** [comment]
besides adding the env was there any change? On my case I could spin up without doing anything. Might as well try on Fedora on VM as I am on Arch @PrimeDominus

**Comment by PrimeDominus** [comment]
@rubiin just got back from a vacation.

I cloned the repo again and tried again on WSL. Not sure if you updated the master branch since I last tried, but the only difference is this time I was able to run `npm install` without having to add `--legacy-peer-deps` or `--force` to complete installing node modules.

I filled out the `./env/.env.dev` file and ran `NODE_ENV=dev make migrate` and once again it couldn't retrieve the values in the file. Also notice the parts I've circled in the screenshot.

![image](https://user-images.githubusercontent.com/31121241/191251283-4dfc9ee3-5187-4449-9f7d-a34d5c1e285a.png)

**Comment by PrimeDominus** [comment]
Why not just have the .env files in the project root?

**Comment by rubiin** [comment]
regarding the `--legacy-peer-deps` its just a npm thing with the nw versions. You can use `yarn` to avoid something like that . Personally I use pnpm but there is a gotcha for `peerDeps`. You have to do `pnpm i --shamefully-hoist=true`

**Comment by rubiin** [comment]
Since you are using the `NODE_ENV=dev`, the file name should be `.env.dev` .  I think you have some typo on the filename as it says `.env` only. Also we are using env folder to reduce the number of files on root to avoid cluttering . Its a personal preference and can be modified if you guys prefer it

**Comment by PrimeDominus** [comment]
You may notice in my last screenshot there's a vscode tab with the filename `.env.dev` which is located in the `env` directory where `.env.example` is located.

**Comment by rubiin** [comment]
Thats strange. Looks like NODE_ENV is not beimg picked. Also can I get a repro so I could see too

**Comment by PrimeDominus** [comment]
> Thats strange. Looks like NODE_ENV is not beimg picked. Also can I get a repro so I could see too

Yes, it's possible NODE_ENV is not getting picked.

I could create a repo for you to reference, but I've changed nothing in the code. As you can see in the screenshot, only `package-lock.json` updated after running `npm install` and the only other thing is I created `./env/.env.dev` by copy-pasting `./env/.env.example` and replacing the values in it (i.e. no keys were changed).

![image](https://user-images.githubusercontent.com/31121241/191522783-3e1b9dd2-73a5-4464-8075-ac9d77e017bc.png)

**Reply by PrimeDominus** [reply]
FYI... We gave up on this repo as is because of the above issues with it.

**Reply by rubiin** [reply]
Thats odd . I talked with another person and he didnt have any issues so far. And also we use this template as a starting point in my office as well , didnt have issues so far

**Reply by rubiin** [reply]
The screenshot is of windows `NODE_ENV=dev npm start` wont work as it is not how we set environment on windows. You may want to use cross-env for that or for windows the readme has `SET NODE_ENV=dev npm run start`

**Comment by mtthsnc** [comment]
Experiencing same issues as @PrimeDominus

**Reply by rubiin** [reply]
what Os do you use. Also mention the commands you ran @omattman

**Reply by rubiin** [reply]
Note: If you are using windows, SET NODE_ENV=dev yarn start or install cross-env or use cross-env NODE_ENV=dev yarn start
